File: internal/registry/registry.go
package registry

import (
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
)

// Agent defines where skills are installed for a supported coding agent.
type Agent struct {
	ID            string `json:"id"`
	DisplayName   string `json:"displayName"`
	GlobalConfig  string `json:"globalConfig"`
	ProjectConfig string `json:"projectConfig"`
	SkillsDir     string `json:"skillsDir"`
}

type Config struct {
	Agents []Agent `json:"agents"`
}

func DefaultAgents() []Agent {
	return []Agent{
		{
			ID:            "codex",
			DisplayName:   "Codex",
			GlobalConfig:  ".codex",
			ProjectConfig: ".codex",
			SkillsDir:     "skills",
		},
		{
			ID:            "claude",
			DisplayName:   "Claude",
			GlobalConfig:  ".claude",
			ProjectConfig: ".claude",
			SkillsDir:     "skills",
		},
		{
			ID:            "amp",
			DisplayName:   "Amp",
			GlobalConfig:  ".config/agents",
			ProjectConfig: ".agents",
			SkillsDir:     "skills",
		},
		{
			ID:            "opencode",
			DisplayName:   "OpenCode",
			GlobalConfig:  ".config/opencode",
			ProjectConfig: ".opencode",
			SkillsDir:     "skill",
		},
		{
			ID:            "goose",
			DisplayName:   "Goose",
			GlobalConfig:  ".config/goose",
			ProjectConfig: ".goose",
			SkillsDir:     "skills",
		},
		{
			ID:            "copilot",
			DisplayName:   "GitHub Copilot",
			GlobalConfig:  "",
			ProjectConfig: ".github",
			SkillsDir:     "skills",
		},
		{
			ID:            "letta",
			DisplayName:   "Letta",
			GlobalConfig:  "",
			ProjectConfig: ".skills",
			SkillsDir:     "",
		},
		{
			ID:            "cursor",
			DisplayName:   "Cursor",
			GlobalConfig:  ".cursor",
			ProjectConfig: ".cursor",
			SkillsDir:     "skills",
		},
	}
}

// LoadRegistry returns default agents plus any overrides from config.
// If the config file is unreadable or invalid, defaults are returned with an error.
func LoadRegistry() ([]Agent, error) {
	defaults := DefaultAgents()
	cfgPaths, err := configPaths()
	if err != nil {
		return defaults, err
	}
	for _, cfgPath := range cfgPaths {
		data, err := os.ReadFile(cfgPath)
		if err != nil {
			if errors.Is(err, os.ErrNotExist) {
				continue
			}
			return defaults, err
		}
		var cfg Config
		if err := json.Unmarshal(data, &cfg); err != nil {
			return defaults, err
		}
		if len(cfg.Agents) == 0 {
			return defaults, nil
		}
		merged := mergeAgents(defaults, cfg.Agents)
		return merged, nil
	}
	return defaults, nil
}

func mergeAgents(defaults, overrides []Agent) []Agent {
	byID := map[string]Agent{}
	order := []string{}
	for _, agent := range defaults {
		if agent.ID == "" {
			continue
		}
		byID[agent.ID] = agent
		order = append(order, agent.ID)
	}
	for _, agent := range overrides {
		if agent.ID == "" {
			continue
		}
		if _, exists := byID[agent.ID]; !exists {
			order = append(order, agent.ID)
		}
		byID[agent.ID] = agent
	}
	merged := make([]Agent, 0, len(byID))
	for _, id := range order {
		if agent, ok := byID[id]; ok {
			merged = append(merged, agent)
		}
	}
	return merged
}

func configPaths() ([]string, error) {
	paths := []string{}
	if configDir, err := os.UserConfigDir(); err == nil && configDir != "" {
		paths = append(paths, filepath.Join(configDir, "knack", "config.json"))
	}
	home, err := os.UserHomeDir()
	if err != nil {
		if len(paths) == 0 {
			return nil, err
		}
		return paths, nil
	}
	legacy := filepath.Join(home, ".config", "knack", "config.json")
	if len(paths) == 0 || paths[0] != legacy {
		paths = append(paths, legacy)
	}
	return paths, nil
}

File: src/registry.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct Agent {
    pub id: String,
    #[serde(rename = "displayName")]
    pub display_name: String,
    #[serde(rename = "globalConfig")]
    pub global_config: String,
    #[serde(rename = "projectConfig")]
    pub project_config: String,
    #[serde(rename = "skillsDir")]
    pub skills_dir: String,
}

pub fn default_agents() -> Vec<Agent> {
    vec![
        Agent {
            id: "codex".to_string(),
            display_name: "Codex".to_string(),
            global_config: ".codex".to_string(),
            project_config: ".codex".to_string(),
            skills_dir: "skills".to_string(),
        },
        Agent {
            id: "claude".to_string(),
            display_name: "Claude".to_string(),
            global_config: ".claude".to_string(),
            project_config: ".claude".to_string(),
            skills_dir: "skills".to_string(),
        },
        Agent {
            id: "amp".to_string(),
            display_name: "Amp".to_string(),
            global_config: ".config/agents".to_string(),
            project_config: ".agents".to_string(),
            skills_dir: "skills".to_string(),
        },
        Agent {
            id: "opencode".to_string(),
            display_name: "OpenCode".to_string(),
            global_config: ".config/opencode".to_string(),
            project_config: ".opencode".to_string(),
            skills_dir: "skill".to_string(),
        },
        Agent {
            id: "goose".to_string(),
            display_name: "Goose".to_string(),
            global_config: ".config/goose".to_string(),
            project_config: ".goose".to_string(),
            skills_dir: "skills".to_string(),
        },
        Agent {
            id: "copilot".to_string(),
            display_name: "GitHub Copilot".to_string(),
            global_config: String::new(),
            project_config: ".github".to_string(),
            skills_dir: "skills".to_string(),
        },
        Agent {
            id: "letta".to_string(),
            display_name: "Letta".to_string(),
            global_config: String::new(),
            project_config: ".skills".to_string(),
            skills_dir: String::new(),
        },
        Agent {
            id: "cursor".to_string(),
            display_name: "Cursor".to_string(),
            global_config: ".cursor".to_string(),
            project_config: ".cursor".to_string(),
            skills_dir: "skills".to_string(),
        },
    ]
}

pub(crate) fn merge_agents(defaults: &[Agent], overrides: &[Agent]) -> Vec<Agent> {
    let mut by_id = std::collections::HashMap::new();
    let mut order = Vec::new();

    for agent in defaults {
        if agent.id.is_empty() {
            continue;
        }
        by_id.insert(agent.id.clone(), agent.clone());
        order.push(agent.id.clone());
    }

    for agent in overrides {
        if agent.id.is_empty() {
            continue;
        }
        if !by_id.contains_key(&agent.id) {
            order.push(agent.id.clone());
        }
        by_id.insert(agent.id.clone(), agent.clone());
    }

    let mut merged = Vec::with_capacity(by_id.len());
    for id in order {
        if let Some(agent) = by_id.get(&id) {
            merged.push(agent.clone());
        }
    }
    merged
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn merge_agents_override() {
        let defaults = vec![
            Agent {
                id: "codex".to_string(),
                display_name: "Codex".to_string(),
                global_config: ".codex".to_string(),
                project_config: ".codex".to_string(),
                skills_dir: "skills".to_string(),
            },
            Agent {
                id: "claude".to_string(),
                display_name: "Claude".to_string(),
                global_config: ".claude".to_string(),
                project_config: ".claude".to_string(),
                skills_dir: "skills".to_string(),
            },
        ];
        let overrides = vec![
            Agent {
                id: "codex".to_string(),
                display_name: "CodexX".to_string(),
                global_config: ".codexx".to_string(),
                project_config: ".codexx".to_string(),
                skills_dir: "skills".to_string(),
            },
            Agent {
                id: "new".to_string(),
                display_name: "New".to_string(),
                global_config: ".new".to_string(),
                project_config: ".new".to_string(),
                skills_dir: "skills".to_string(),
            },
        ];

        let merged = merge_agents(&defaults, &overrides);
        assert_eq!(merged.len(), 3);
        assert_eq!(merged[0].display_name, "CodexX");
        assert_eq!(merged[2].id, "new");
    }
}
